### PR TITLE
Allow custom amount field in Donor Dasboard Subscription Management to be emptied

### LINKED ISSUE #5896: Custom Amount field on Donor Dashboard Manage Subscription should allow emptying the input
## Details

Right now, a user is not able to clear this input. It will always at least display the minimum value. While this might seem like it's the intended behaviour, it's actually overly zealous and creates a bad user experience, since user input in this type of field is at points partial as they are typing in the value.

## Expected Behavior

The user should be allowed to clear the input. I don’t even think this field should be correcting the value on change or on blur. One reason is that it is currency and that is deceiving. We do not want to be responsible for changing an intended amount. It would be better to explicitly notify them that the intended (or unintended) amount they enter was invalid, instead of changing it as they type it in.

## Steps to Reproduce

1. Create a recurring donation.
2. Go to the Donor Dashboard.
3. In "Recurring Donations" find the donation.
4. In the subscription management, try clearing out the amount field (for custom amount).

## Additional Context

The problem for this is all in the frontend. Looks like either the `AmountControl` or the `CurrencyControl` are the culprit.

## Acceptance Criteria

- [x] Allow empty values in for `AmountControl` or `CurrencyControl` (i.e. do not zealously mask the input).
- [x] Users can adequately change their subscription amount.
- [x] If a value less than the minimum or more than the maximum is provided, either frontend or backend validation notifies the user they need to adjust the value.

### PR DESCRIPTION
**Pull request by knowler**
Resolves #5896

## Description

In the subscription management view of the donor dashboard, there is an input that appears when the user chooses "Custom Amount." This input previously would replace its empty state with the minimum allowed value. This seems like it is desirable, however, it is not. In fact, users without the knack of filling out forms on the Internet will find it impossible to change the value to an amount where the most significant digit is less than the most significant digit in the minimum value (e.g. the minimum is $5 and the user wants to update the value to $12, but needs to change it from $15, once they backspace once, so that the input $reads 1, the input will think the value is less than $5 and change it to $5, since $1 is less than $5, even though they intend to write $12).

This PR resolves this issue by not setting minimum and maximum values when the value in the input is outside of them. Instead, we simply validate the input on blur and provide the user the appropriate feedback they need to enter an acceptable amount.

The existing logic in these components was so hard to follow and, since I had redefined the functionality, I completely scrapped what was there. The implementation that I needed was far simpler and I reduced the complexity further by normalizing the amount, the minimum, and the maximum _before_ giving them to the controls. This means that the underlying controls work in a more controlled manner.

This also gets rid of the `CurrencyControl` component. This was the only view that was using it, so I think it was a premature abstraction. We can better think through a re-usable version in the future.

## Affects

- Subscription Management view of the Donor Dashboard.

## Visuals

https://user-images.githubusercontent.com/6908001/135361310-2ba270b2-3827-4d66-b7f8-29b0477fe0e4.mp4

## Testing Instructions

Requirements:

- Recurring addon is installed and activated.
- Make sure you are using the Stripe gateway.

Instructions:

1. Set up a form that allows recurring donations.
2. Make a recurring donation.
3. View the donation dashboard for the donor (`/donor-dashboard`).
4. Go to "Recurring Donations."
4. Press "Manage Subscription" on the recurring donation.
5. Set the select to "Custom Amount."
6. Play around with setting the value correctly and incorrectly.

In general you might want to test the whole subscription management feature out since I changed a fair bit of code there.

## Pre-review Checklist

- [x] Acceptance criteria satisfied and marked in related issue
- [ ] Relevant `@since` tags included in DocBlocks
- [x] Changes logged to the `Unreleased` section of `CHANGELOG.md`
- [ ] Includes unit and/or end-to-end tests
- [x] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

### CODE REVIEW

**Comment by knowler** [comment]
This is ready for review.

@DevinWalker I implemented a fairly simple validation error message, I just thought I should tag you in case you had any feedback on it.

File: src/DonorDashboards/resources/js/app/components/subscription-manager/amount-control/index.js [hunk 1]
```diff
@@ -1,88 +1,156 @@
...
-			} );
-		}
+/**
+ * Quick and cheap knock-off of the classnames package.
+ * Do not re-use. Just use the real package.
+ *
+ * @param {...any} className
+ * @return {string}
+ */
+const cx = ( ...classNames ) => classNames.filter( Boolean ).join( ' ' );
```
Comment (alaca) [inline src/DonorDashboards/resources/js/app/components/subscription-manager/amount-control/index.js:29]: Cool.

**Review by alaca — APPROVED** [review]
Looks good and is much more optimized now. Great work, @knowler

**Review by ravinderk — CHANGES_REQUESTED** [review]
@knowler Did we test this with currency switcher?

**Comment by knowler** [comment]
@ravinderk I tested this with Euros (which uses comma as separator) and it seems to work. If we need more extensive testing with currency switcher, then I will need some help as it took me a while to figure out a currency that worked with the payment gateway I was using (Stripe Credit Card).
